src/cockpit/channels/filesystem.py

```diff
@@ -158,6 +160,64 @@ def do_yield_data(self, options: JsonObject) -> Generator[bytes, None, JsonObjec
             raise ChannelError('internal-error', message=str(exc)) from exc
 
 
+class FSReplaceAttrs:
+    supported_attrs = ['uid', 'gid', 'owner', 'group']
+
+    def __init__(self, value: JsonObject) -> None:
+        # Any unknown keys throw an error
+        unsupported_attrs = list(value.keys() - self.supported_attrs)
+        if unsupported_attrs:
+            raise ChannelError('protocol-error',
+                               message=f'"attrs" contains unsupported key(s) {unsupported_attrs}') from None
+
+        self.uid = get_int(value, 'uid', None)
+        self.gid = get_int(value, 'gid', None)
+
+        if self.uid is None and self.gid is not None:
+            raise ChannelError('protocol-error', message='cannot provide a gid without an uid')
```

Comment:
Hmm, why not?

Reply:
> If a colon but no group name follows the user name, that user is made  the  owner  of  the files and the group of the files is changed to that user's login group.

(from the man page). TIL!

However, we are not calling `chown(1)` here, but the fchown() syscall, which doesn't have that magic. So unrelated to this PR, but still useful to know -- thanks!

Reply:
Yup, let's leave this out!